src/treq/test/local_httpbin/parent.py

```diff
@@ -0,0 +1,176 @@
+
+from .shared import _HTTPBinDescription
+
+
+class _HTTPBinServerProcessProtocol(basic.LineOnlyReceiver):
+    """
+    Manage the lifecycle of an ``httpbin`` process.
+    """
+    delimiter = os.linesep.encode('ascii')
+
+    def __init__(self, all_data_received, terminated):
+        """
+        Manage the lifecycle of an ``httpbin`` process.
+
+        :param all_data_received: A Deferred that will be called back
+            with an :py:class:`_HTTPBinDescription` object
+        :type all_data_received: :py:class:`Deferred`
+
+        :param terminated: A Deferred that will be called back when
+            the process has ended.
+        :type terminated: :py:class:`Deferred`
+        """
+        self._all_data_received = all_data_received
+        self._received = False
+        self._terminated = terminated
+
+    def lineReceived(self, line):
+        if self._received:
+            raise RuntimeError("Unexpected line: {!r}".format(line))
+        description = _HTTPBinDescription.from_json_bytes(line)
+
+        self._received = True
+
+        # Remove readers and writers that leave the reactor in a dirty
+        # state after a test.
+        self.transport.closeStdin()
+        self.transport.closeStdout()
+        self.transport.closeStderr()
+
+        self._all_data_received.callback(description)
+
+    def connectionLost(self, reason):
+        if not self._received:
+            self._all_data_received.errback(reason)
+        self._terminated.errback(reason)
+
+
+@attr.s
+class _HTTPBinProcess(object):
+    """
+    Manage an ``httpbin`` server process.
+
+    :ivar _all_data_received: See
+        :py:attr:`_HTTPBinServerProcessProtocol.all_data_received`
+    :ivar _terminated: See
+        :py:attr:`_HTTPBinServerProcessProtocol.terminated`
+    """
+    _https = attr.ib()
+
+    _error_log_path = attr.ib(default='httpbin-server-error.log')
+
+    _all_data_received = attr.ib(init=False, default=attr.Factory(Deferred))
+    _terminated = attr.ib(init=False, default=attr.Factory(Deferred))
+
+    _process = attr.ib(init=False, default=None)
+    _process_description = attr.ib(init=False, default=None)
+
+    _open = staticmethod(open)
+
+    def _spawn_httpbin_process(self, reactor):
+        """
+        Spawn an ``httpbin`` process, returning a :py:class:`Deferred`
+        that fires with the process transport and result.
+        """
+        server = _HTTPBinServerProcessProtocol(
+            all_data_received=self._all_data_received,
+            terminated=self._terminated
+        )
+
+        argv = [
+            sys.executable,
+            '-m',
+            'treq.test.local_httpbin.child',
+        ]
+
+        if self._https:
+            argv.append('--https')
+
+        with self._open(self._error_log_path, 'wb') as error_log:
+            endpoint = endpoints.ProcessEndpoint(
+                reactor,
+                sys.executable,
+                argv,
+                childFDs={
```

Comment:
OK, so now I remember why I did this: the parent closes its standard error file descriptor immediately to avoid a dirty reactor failure.

By redirecting standard error to a file before this happens, I hoped to capture any useful log messages without re-registering that error file descriptor.  I'm going to re-add this code.